pymc_marketing/mmm/delayed_saturated_mmm.py

```diff
@@ -63,80 +55,128 @@ def __init__(
         )
 
-    def _preprocess_channel_prior(self) -> TensorVariable:
-        return (
-            pm.HalfNormal.dist(sigma=2, shape=len(self.channel_columns))
-            if self.channel_prior is None
-            else change_dist_size(
-                dist=self.channel_prior, new_size=len(self.channel_columns)
-            )
-        )
+    @property
+    def default_sampler_config(self) -> Dict:
+        return {"progressbar": True, "random_seed": 1234}
 
-    def build_model(
-        self,
-        data: pd.DataFrame,
-        adstock_max_lag: int = 4,
+    @property
+    def output_var(self):
+        return "y"
+
+    def generate_and_preprocess_model_data(
+        self, X: Union[pd.DataFrame, pd.Series], y: pd.Series
     ) -> None:
-        date_data = data[self.date_column]
-        target_data = data[self.target_column]
-        channel_data = data[self.channel_columns]
+        """
+        Applies preprocessing to the data before fitting the model.
+        if validate is True, it will check if the data is valid for the model.
+        sets self.model_coords based on provided dataset
 
+        Parameters
+        ----------
+        X : array, shape (n_obs, n_features)
+        y : array, shape (n_obs,)
+        """
+        date_data = X[self.date_column]
+        channel_data = X[self.channel_columns]
         coords: Dict[str, Any] = {
             "date": date_data,
-            "channel": channel_data.columns,
+            "channel": self.channel_columns,
         }
 
+        new_X_dict = {
+            self.date_column: date_data,
+        }
+        X_data = pd.DataFrame.from_dict(new_X_dict)
+        X_data = pd.concat([X_data, channel_data], axis=1)
+        control_data: Optional[Union[pd.DataFrame, pd.Series]] = None
         if self.control_columns is not None:
-            control_data: Optional[pd.DataFrame] = data[self.control_columns]
-            coords["control"] = data[self.control_columns].columns
-        else:
-            control_data = None
+            control_data = X[self.control_columns]
+            coords["control"] = self.control_columns
+            X_data = pd.concat([X_data, control_data], axis=1)
 
+        fourier_features: Optional[pd.DataFrame] = None
         if self.yearly_seasonality is not None:
-            fourier_features = self._get_fourier_models_data()
+            fourier_features = self._get_fourier_models_data(X=X)
+            self.fourier_columns = fourier_features.columns
             coords["fourier_mode"] = fourier_features.columns.to_numpy()
+            X_data = pd.concat([X_data, fourier_features], axis=1)
+
+        self.model_coords = coords
+        if self.validate_data:
+            self.validate("X", X_data)
+            self.validate("y", y)
+        self.preprocessed_data: Dict[str, Union[pd.DataFrame, pd.Series]] = {
+            "X": self.preprocess("X", X_data.copy()),
+            "y": self.preprocess("y", y.copy()),
```

Comment:
Why the copy?

Reply:
the original implementation was working on a copy of the data, so I kept it

Reply:
Okay, seems wasteful, but if it's not your change, no need to discuss it here

Reply:
I checked with tests and everything still works as expected without the copies, so I removed it.